pkg/protocol/block_dispatcher.go

```diff
@@ -0,0 +1,345 @@
+//
+// This is the case if the block is more than a warp sync threshold ahead of the latest commitment while also committing
+// to a new slot that can be warp synced.
+func (b *BlockDispatcher) inWarpSyncRange(engine *engine.Engine, block *model.Block) bool {
+	if engine.BlockRequester.HasTicker(block.ID()) {
+		return false
+	}
+
+	slotCommitmentID := block.ProtocolBlock().SlotCommitmentID
+	latestCommitmentIndex := engine.Storage.Settings().LatestCommitment().Index()
+	maxCommittableAge := engine.APIForSlot(slotCommitmentID.Index()).ProtocolParameters().MaxCommittableAge()
+
+	return block.ID().Index() > latestCommitmentIndex+maxCommittableAge
+}
+
+// warpSync triggers warp sync from the latest committed slot up to the warpsync window.
+func (b *BlockDispatcher) warpSync(e *engine.Engine, chainCommitment *chainmanager.ChainCommitment) {
+	if e == nil || chainCommitment == nil {
+		return
+	}
+
+	chain := chainCommitment.Chain()
+	maxCommittableAge := e.APIForSlot(chainCommitment.Commitment().Index()).ProtocolParameters().MaxCommittableAge()
+	warpSyncWindowSize := lo.Cond(maxCommittableAge > b.optWarpSyncWindowSize, maxCommittableAge, b.optWarpSyncWindowSize)
+	latestCommitmentIndex := e.Storage.Settings().LatestCommitment().Index()
+
+	for slotToWarpSync := latestCommitmentIndex + 1; slotToWarpSync <= latestCommitmentIndex+warpSyncWindowSize; slotToWarpSync++ {
+		if commitmentToSync := chain.Commitment(slotToWarpSync); commitmentToSync != nil && !b.processedWarpSyncRequests.Has(commitmentToSync.ID()) {
+			b.pendingWarpSyncRequests.StartTicker(commitmentToSync.ID())
+		}
+	}
+}
+
+// injectUnsolidCommitmentBlocks injects the unsolid blocks for the given commitment ID into the correct engine
+// instance.
+func (b *BlockDispatcher) injectUnsolidCommitmentBlocks(id iotago.CommitmentID) {
+	for _, tuple := range b.unsolidCommitmentBlocks.GetValues(id) {
+		b.protocol.HandleError(b.Dispatch(tuple.A, tuple.B))
+	}
+}
+
+// targetEngine returns the engine instance that should be used for the given commitment.
+func (b *BlockDispatcher) targetEngine(commitment *chainmanager.ChainCommitment) *engine.Engine {
+	if chain := commitment.Chain(); chain != nil {
+		chainID := chain.ForkingPoint.Commitment().ID()
+
+		if engine := b.protocol.MainEngineInstance(); engine.ChainID() == chainID {
+			return engine
+		}
+
+		if engine := b.protocol.CandidateEngineInstance(); engine != nil && engine.ChainID() == chainID {
+			return engine
+		}
+	}
+
+	return nil
+}
+
+// monitorLatestEngineCommitment monitors the latest commitment of the given engine instance and triggers a warp sync if
+// necessary.
+func (b *BlockDispatcher) monitorLatestEngineCommitment(engineInstance *engine.Engine) {
+	engineInstance.HookStopped(engineInstance.Events.Notarization.LatestCommitmentUpdated.Hook(func(commitment *model.Commitment) {
+		if chainCommitment, exists := b.protocol.ChainManager.Commitment(commitment.ID()); exists {
```

Comment:
Imagine you are syncing a new node.

You see a few new commitments, and then you solidify them. Once you have a solid commitment chain you start to warpsync the first n epochs.

Every time you commit the engine, you need to advance your warpsync window, to now request the next epoch so you start to slowly catch up. This is what happens here - as soon as we commit we adjust our warpsync window and queue the new slots that should be requested.

The case that you described where you are close to the tip of the chain and you i.e. see a commitment slightly earlier than you commit yourself should actually not cause a warpsync because there was a cut off condition in "warpSync" that prevented this from happening when you were close to the tip of the chain.

It was also called "warpSyncIfNecessary". Andrea seems to have removed this line of code and I am waiting for an answer for why he did that as you are right that we would currently warpsync epochs that we are about to commit ourselves.

Reply:
After talking to Andrea, we re-added a cut off condition for the case where we are close to the tip of the chain to prevent us from warp syncing slots that we are about to commit ourselves.